### PR TITLE
Switch parser to argparse

### LINKED ISSUE #1257: switch to argparse for argument parsing

### PR DESCRIPTION
**Pull request by ddfisher**
I had this partially written, so I thought it was worth taking an hour to get it in.  From here, we can start simplifying build flags/Options and other parts of `main.py`.

Fixes #1257.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Many tests fail with `AttributeError: 'Namespace' object has no
attribute 'fast_parse'` :-(

**Comment by ddfisher** [comment]
Ah shoot I know what that is. Silly error. :/ Will fix when I get home.

File: mypy/main.py [hunk 1]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
             module to run as script (or None),
             parsed flags)
     """
-    # TODO: Rewrite using argparse.
+
+    footer = textwrap.dedent(
```
Comment (gvanrossum) [inline mypy/main.py]: I don't think it's worth importing textwrap just so you can write this indented. Maybe make it a global constant FOOTER?

File: mypy/main.py [hunk 2]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+    def parse_version(v):
+        version_components = v.split(".")[0:2]
+        if len(version_components) != 2:
+            raise argparse.ArgumentTypeError(
+                "Invalid python version '{}' (expected format: 'x.y')".format(v))
+        if not all(item.isdigit() for item in version_components):
+            raise argparse.ArgumentTypeError("Found non-digit in python version: '{}'".format(v))
+        return (int(version_components[0]), int(version_components[1]))
+
+    parser.add_argument('-v', '--verbose', action='store_true', help="more verbose messages")
```
Comment (gvanrossum) [inline mypy/main.py]: Can you make this a counter? I've changed things in build.py so that trace() calls print only with `-v -v` while log() prints with `-v`.

File: mypy/main.py [hunk 3]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+    parser.add_argument('--version', action='version', version='%(prog)s ' + __version__)
```
Comment (gvanrossum) [inline mypy/main.py]: Can you add -V as an alternative here? (It's a common convention.)

File: mypy/main.py [hunk 4]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+    parser.add_argument('--python-version', type=parse_version, metavar='x.y',
+                        help='use Python x.y')
+    parser.add_argument('--py2', action='store_true', help="use Python 2 mode")
```
Comment (gvanrossum) [inline mypy/main.py]: Couldn't the action here be made to store (2, 7) in python version?

File: mypy/main.py [hunk 5]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+
+    footer = textwrap.dedent(
+        """environment variables:
+        MYPYPATH     additional module search path"""
+    )
+
+    parser = argparse.ArgumentParser(prog='mypy', epilog=footer,
+                                     formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    def parse_version(v):
```
Comment (gvanrossum) [inline mypy/main.py]: Honestly here I would use a regex, e.g. `r'\A(\d)\.(\d+)\Z'`.

File: mypy/main.py [hunk 6]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+        parser.error('Python version 2 (or --py2) specified, '
+                     'but --use-python-path will search in sys.path of Python 3')
+
+    if args.fast_parser and (args.py2 or
+                             args.python_version and args.python_version[0] == 2):
+        parser.error('The experimental fast parser is only compatible with Python 3, '
+                     'but Python 2 specified.')
+
+    # Set options.
     options = Options()
```
Comment (gvanrossum) [inline mypy/main.py]: It's a shame that we still need the Options class, but I understand it's typed which is harder for the Namespace returned by parse_args().

File: mypy/main.py [hunk 7]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+
+    if args.disallow_untyped_defs:
+        options.build_flags.append(build.DISALLOW_UNTYPED_DEFS)
+
+    # experimental
+    if args.fast_parse:
+        options.build_flags.append(build.FAST_PARSER)
+
+    # Set reports.
+    for flag, val in vars(args).items():
```
Comment (gvanrossum) [inline mypy/main.py]: Can you do this without resorting to vars()? It feels unnecessarily dynamic. Four similar 2-line blocks wouldn't be so bad IMO.

File: mypy/main.py [hunk 8]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+        if flag.endswith('_report') and val is not None:
+            report_type = flag[:-7].replace('_', '-')
+            report_dir = val
+            options.report_dirs[report_type] = report_dir
+
+    # Set target.
```
Comment (gvanrossum) [inline mypy/main.py]: For -m, --package, -c, are you checking that there are no spurious extra arguments? mypy ignored those until I added explicit checks for them recently, presumably because when it could still run your code those would become the sys.argv of the execution. But now they just confuse users who accidentally pass them in.

File: mypy/main.py [hunk 9]
```diff
@@ -102,120 +104,152 @@ def process_options(args: List[str]) -> Tuple[List[BuildSource], Options]:
...
+        return [BuildSource(None, None, args.command)], options
+    else:
+        targets = []
+        for f in args.files:
+            if f.endswith(PY_EXTENSIONS):
+                targets.append(BuildSource(f, crawl_up(f)[1], None))
+            elif os.path.isdir(f):
+                targets.extend(expand_dir(f))
+            else:
+                targets.append(BuildSource(f, None, None))
```
Comment (gvanrossum) [inline mypy/main.py]: Would be nice to check there's at most one of these (I guess this is for scripts), since they all get the module name `__main__` and you can't have duplicate module names. There's a check for that in build.py but it's a bit ugly so late in the game (it used to be an assert :-).

File: mypy/main.py (hunk 2 above)
Comment (ddfisher) [inline mypy/main.py]: Done.

File: mypy/main.py (hunk 3 above)
Comment (ddfisher) [inline mypy/main.py]: Added.

File: mypy/main.py (hunk 4 above)
Comment (ddfisher) [inline mypy/main.py]: Yep, that's better! I forgot to go back and clean it up after realizing that's the only thing that `--py2` did.  Changed.

File: mypy/main.py (hunk 5 above)
Comment (ddfisher) [inline mypy/main.py]: This is old code I moved around.  I agree a regex is nicer -- changed!

File: mypy/main.py (hunk 6 above)
Comment (ddfisher) [inline mypy/main.py]: I think we'll be able to do something nicer here in the future (like maybe assigning directly to an Options with a custom Action), but I wanted to get the basic version working first.

File: mypy/main.py (hunk 7 above)
Comment (ddfisher) [inline mypy/main.py]: I think it would be a bit unfortunate if we had to list out all the reports twice (but I'm not strongly opposed).  I only used `vars()` here because I was explicitly told to do this in the argparse documentation! https://docs.python.org/3/library/argparse.html#argparse.Namespace

File: mypy/main.py (hunk 8 above)
Comment (ddfisher) [inline mypy/main.py]: Yep!  This happens here: https://github.com/python/mypy/pull/1295/files#diff-b5e3909c86f7834f70af20bb848acec3R170

File: mypy/main.py (hunk 9 above)
Comment (ddfisher) [inline mypy/main.py]: Could we add that in a different PR?  This is just old code that I moved around, and that change seems a bit unrelated (and I also only have a vague understanding of what's going on here).

File: mypy/main.py (hunk 1 above)
Comment (ddfisher) [inline mypy/main.py]: Sure.  `dedent` has always felt a bit unfortunate to me.

**Comment by gvanrossum** [comment]
There's still one failing test. I will drop the remaining requests!

**Comment by ddfisher** [comment]
It looks like f119547afa425f553a0f9665d0c4c38469216e1a broke the one failing test (which is in a completely unrelated area).
